### PR TITLE
fix phpunit 11 config

### PR DESCRIPTION
**Pull request by Kanti**
As described in #2001 with phpunit 11 infection is broken.
I added the test from #2001 and fixed the configuration by:
- replacing `cacheResultFile`with `cacheDirectory`for every phpunit installation >=11
- replacing `stopOnFailure`with `stopOnDefect` to speed up the infection time by 4x (on my project)

TODOs:
- [x]  find out how we can skip the PHPUnit11 e2e folder in Github Actions using PHP <8.2 (Maybe someone can help me here as I'm not familiar with this test setup.)

### CODE REVIEW

File: src/TestFramework/PhpUnit/Config/XmlConfigurationManipulator.php [hunk 1]
```diff
@@ -117,13 +126,16 @@ public function deactivateStderrRedirection(SafeDOMXPath $xPath): void
...
 
-    public function setStopOnFailure(SafeDOMXPath $xPath): void
+    public function setStopOnFailure(string $version, SafeDOMXPath $xPath): void
     {
-        $this->setAttributeValue(
-            $xPath,
-            'stopOnFailure',
-            'true',
-        );
+        # with phpunit 10.0  we need to use stopOnDefect instead: https://github.com/sebastianbergmann/phpunit/blob/10.0.0/ChangeLog-10.0.md
```
Comment (Kanti) [inline src/TestFramework/PhpUnit/Config/XmlConfigurationManipulator.php:131]:
```suggestion
        // with phpunit 10.0  we need to use stopOnDefect instead: https://github.com/sebastianbergmann/phpunit/blob/10.0.0/ChangeLog-10.0.md
```

**Comment by maks-rafalko** [comment]
> * Maybe someone can help me here as I'm not familiar with this test setup.

one possible way is to add `run_tests.bash` to your e2e test folder and there you can check (pseudocode)

```
if %php version < 8.2%
   exit 0
endif
```

Regarding `cacheResultFilewith` - I will need to dig into it, will try to allocate time in the nearest days, but it won't be fast as I quite limited these days.

**Comment by maks-rafalko** [comment]
> replacing `stopOnFailure`with `stopOnDefect` to speed up the infection time by 4x (on my project)

that's crazy that PHPUnit doesn't fail with incorrect flag usage :( Infection is so much faster when tests execution stops at the first failure rather than executing all the tests. Thanks for fixing it, this is a major bugfix

One question though: how `cacheDirectory` and `stopOnDefect` is related to `testdox` issue you are referencing?

File: src/TestFramework/PhpUnit/Config/XmlConfigurationManipulator.php [hunk 2]
```diff
@@ -117,13 +126,16 @@ public function deactivateStderrRedirection(SafeDOMXPath $xPath): void
         $this->setAttributeValue($xPath, 'stderr', 'false');
     }
 
-    public function setStopOnFailure(SafeDOMXPath $xPath): void
+    public function setStopOnFailure(string $version, SafeDOMXPath $xPath): void
```
Comment (maks-rafalko) [inline src/TestFramework/PhpUnit/Config/XmlConfigurationManipulator.php:129]: minor: what do you think about renaming to `setStopOnFailureOrDefect`?

File: tests/e2e/PHPUnit11/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Title
+
+* Link to github ticket if relevant
```
Comment (maks-rafalko) [inline tests/e2e/PHPUnit11/README.md:3]: Please add link to the GH issue/pr and update summary please

**Comment by maks-rafalko** [comment]
> speed up the infection time by 4x (on my project)

did you analyze your project's Infection log file to understand why there is such a huge difference? I've tested `master`  branch to run Infection on itself and your branch, there is no difference, at least from the first look.

**Comment by Kanti** [comment]
The 4x performance gain is because of my testing structure. It will not effect all projects by that much.

I looked into it and found out that I have mutations that are covered by hundreds of test Cases.
And some mutations trigger an error, that error will not trigger phpunit to stop running the tests. So phpunit will run all the hundreds of tests even if the first one has errored.

With the `stopOnDefect` it will stop after the first error. As well as the first failure.

If you don't have that many test cases that are covering a mutation or the mutation does not trigger an error, you won't have that much of a performance gain.

I thought that the `testdox` was the problem, but it was not. It was the `cacheResultFile` all along.
As I worked on that I also found that `setStopOnFailureOrDefect` needs a change, it was more a coincidence.

**Comment by maks-rafalko** [comment]
ok, thanks for the explanation, now it's much clear. 

> And some mutations trigger an error, that error will not trigger phpunit to stop running the tests.

could you please provide an example of the original code and mutated code that triggers error and stop only with `stopOnDefect` but not with `stopOnFailure`? 

I would like to play with it and see how it affects other users.

**Comment by Kanti** [comment]
I looked into my project and found one problem:
If the type dose not match it is an error and not failure.
I recreated a fictitious test case here: https://infection-php.dev/r/1l4r

As you can see, it runs all 2000 test cases instead of stopping after the first:
grafik

**Comment by maks-rafalko** [comment]
Ok, regarding `cacheResultFile` being removed from PHPUnit 11, just for history and search purposes, here is what was the issue:

```
  1) Test results may not be as expected because the XML configuration file did not pass validation:
  
    Line 2:
    - Element 'phpunit', attribute 'cacheResultFile': The attribute 'cacheResultFile' is not allowed.
  
  
  No tests executed!
```

so this PR solves correctly.

Thank you so much @Kanti for your work on fixing and adding improvements (`stopOnDefect`), this is very helpful and useful.

**Comment by maks-rafalko** [comment]
Released https://github.com/infection/infection/releases/tag/0.29.8

**Comment by maks-rafalko** [comment]
Updated Playground so it clearly visible how this PR improved Infection execution:

- before https://infection-php.dev/r/1l4r
- after https://infection-php.dev/r/751k

**Pull request merged**